pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -391,173 +415,167 @@ func validateParamArrayIndex(ctx context.Context, params []v1beta1.Param, spec *
 			if p.Default.Type == v1beta1.ParamTypeArray {
 				for _, pattern := range patterns {
 					for i := 0; i < len(p.Default.ArrayVal); i++ {
-						arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
+						arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Default.ArrayVal)
 					}
 				}
 			}
 		}
 	}
 
-	// Collect array params lengths from pipeline
+	// Collect array params lengths from params
 	for _, p := range params {
 		if p.Value.Type == v1beta1.ParamTypeArray {
 			for _, pattern := range patterns {
 				for i := 0; i < len(p.Value.ArrayVal); i++ {
-					arrayParams[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
+					arrayParamsLengths[fmt.Sprintf(pattern, p.Name)] = len(p.Value.ArrayVal)
 				}
 			}
 		}
 	}
+	return arrayParamsLengths
+}
 
+// ValidateOutofBoundArrayParams validates if the array indexing params are out of bound
+func ValidateOutofBoundArrayParams(arrayIndexingParams []string, arrayParamsLengths map[string]int) error {
 	outofBoundParams := sets.String{}
-
-	// Validate array param in steps fields.
-	validateStepsParamArrayIndexing(spec.Steps, arrayParams, &outofBoundParams)
-
-	// Validate array param in StepTemplate fields.
-	validateStepsTemplateParamArrayIndexing(spec.StepTemplate, arrayParams, &outofBoundParams)
-
-	// Validate array param in build's volumes
-	validateVolumesParamArrayIndexing(spec.Volumes, arrayParams, &outofBoundParams)
-
-	for _, v := range spec.Workspaces {
-		extractParamIndex(v.MountPath, arrayParams, &outofBoundParams)
+	for _, val := range arrayIndexingParams {
+		indexString := substitution.ExtractIndexString(val)
+		idx, _ := substitution.ExtractIndex(indexString)
+		v := substitution.TrimArrayIndex(val)
+		if paramLength, ok := arrayParamsLengths[v]; ok {
+			if idx >= paramLength {
+				outofBoundParams.Insert(val)
+			}
+		}
 	}
-
-	validateSidecarsParamArrayIndexing(spec.Sidecars, arrayParams, &outofBoundParams)
-
 	if outofBoundParams.Len() > 0 {
 		return fmt.Errorf("non-existent param references:%v", outofBoundParams.List())
 	}
-
 	return nil
 }
 
-func extractParamIndex(paramReference string, arrayParams map[string]int, outofBoundParams *sets.String) {
+// ExtractArrayIndexingParamRefs extract the array indexing references
+func ExtractArrayIndexingParamRefs(paramReference string, arrayIndexingParams *[]string) {
```

Comment:
a pointer to a slice seems a bit weird, because you probably don't want to distinguish between a nil slice and a nil pointer to a slice. any reason to not use a regular slice?

Reply:
I didn't realize this function modifies the input! I prefer having functions that don't modify their inputs (and if they do modify their inputs, it should be very clear from the docstring). This can cause bugs like the one fixed in https://github.com/tektoncd/pipeline/pull/6004. For that reason and the one I mentioned in the original comment, I think returning a slice is better.

Reply:
Sure! I'm changing it now. Thanks!

Reply:
updated